File: src_original/extras-solid/libdevice.cpp
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#define WORKAROUND_INNER_CLASS_DEFINITION_Solid__Device
#define WORKAROUND_INNER_CLASS_DEFINITION_Solid__DeviceInterface
#define WORKAROUND_INNER_CLASS_DEFINITION_Solid__Predicate
#include <device.h>
#include "libdevice.h"
#include "libdevice.hxx"

Solid__Device* Solid__Device_new() {
    return new Solid::Device();
}

Solid__Device* Solid__Device_new2(const Solid__Device* device) {
    return new Solid::Device(*device);
}

Solid__Device* Solid__Device_new3(const libqt_string udi) {
    QString udi_QString = QString::fromUtf8(udi.data, udi.len);
    return new Solid::Device(udi_QString);
}

libqt_list /* of Solid__Device* */ Solid__Device_AllDevices() {
    QList<Solid::Device> _ret = Solid::Device::allDevices();
    // Convert QList<> from C++ memory to manually-managed C memory
    Solid__Device** _arr = static_cast<Solid__Device**>(malloc(sizeof(Solid__Device*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new Solid::Device(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of Solid__Device* */ Solid__Device_ListFromType(const int* typeVal) {
    QList<Solid::Device> _ret = Solid::Device::listFromType((const Solid::DeviceInterface::Type&)(*typeVal));
    // Convert QList<> from C++ memory to manually-managed C memory
    Solid__Device** _arr = static_cast<Solid__Device**>(malloc(sizeof(Solid__Device*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new Solid::Device(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of Solid__Device* */ Solid__Device_ListFromQuery(const Solid__Predicate* predicate) {
    QList<Solid::Device> _ret = Solid::Device::listFromQuery(*predicate);
    // Convert QList<> from C++ memory to manually-managed C memory
    Solid__Device** _arr = static_cast<Solid__Device**>(malloc(sizeof(Solid__Device*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new Solid::Device(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of Solid__Device* */ Solid__Device_ListFromQuery2(const libqt_string predicate) {
    QString predicate_QString = QString::fromUtf8(predicate.data, predicate.len);
    QList<Solid::Device> _ret = Solid::Device::listFromQuery(predicate_QString);
    // Convert QList<> from C++ memory to manually-managed C memory
    Solid__Device** _arr = static_cast<Solid__Device**>(malloc(sizeof(Solid__Device*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new Solid::Device(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

Solid__Device* Solid__Device_StorageAccessFromPath(const libqt_string path) {
    QString path_QString = QString::fromUtf8(path.data, path.len);
    return new Solid::Device(Solid::Device::storageAccessFromPath(path_QString));
}

void Solid__Device_OperatorAssign(Solid__Device* self, const Solid__Device* device) {
    self->operator=(*device);
}

bool Solid__Device_IsValid(const Solid__Device* self) {
    return self->isValid();
}

libqt_string Solid__Device_Udi(const Solid__Device* self) {
    QString _ret = self->udi();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Solid__Device_ParentUdi(const Solid__Device* self) {
    QString _ret = self->parentUdi();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

Solid__Device* Solid__Device_Parent(const Solid__Device* self) {
    return new Solid::Device(self->parent());
}

libqt_string Solid__Device_Vendor(const Solid__Device* self) {
    QString _ret = self->vendor();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Solid__Device_Product(const Solid__Device* self) {
    QString _ret = self->product();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Solid__Device_Icon(const Solid__Device* self) {
    QString _ret = self->icon();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of libqt_string */ Solid__Device_Emblems(const Solid__Device* self) {
    QList<QString> _ret = self->emblems();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string Solid__Device_DisplayName(const Solid__Device* self) {
    QString _ret = self->displayName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Solid__Device_Description(const Solid__Device* self) {
    QString _ret = self->description();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool Solid__Device_IsDeviceInterface(const Solid__Device* self, const int* typeVal) {
    return self->isDeviceInterface((const Solid::DeviceInterface::Type&)(*typeVal));
}

Solid__DeviceInterface* Solid__Device_AsDeviceInterface(Solid__Device* self, const int* typeVal) {
    return self->asDeviceInterface((const Solid::DeviceInterface::Type&)(*typeVal));
}

Solid__DeviceInterface* Solid__Device_AsDeviceInterface2(const Solid__Device* self, const int* typeVal) {
    return (Solid__DeviceInterface*)self->asDeviceInterface((const Solid::DeviceInterface::Type&)(*typeVal));
}

libqt_list /* of Solid__Device* */ Solid__Device_ListFromType2(const int* typeVal, const libqt_string parentUdi) {
    QString parentUdi_QString = QString::fromUtf8(parentUdi.data, parentUdi.len);
    QList<Solid::Device> _ret = Solid::Device::listFromType((const Solid::DeviceInterface::Type&)(*typeVal), parentUdi_QString);
    // Convert QList<> from C++ memory to manually-managed C memory
    Solid__Device** _arr = static_cast<Solid__Device**>(malloc(sizeof(Solid__Device*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new Solid::Device(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of Solid__Device* */ Solid__Device_ListFromQuery22(const Solid__Predicate* predicate, const libqt_string parentUdi) {
    QString parentUdi_QString = QString::fromUtf8(parentUdi.data, parentUdi.len);
    QList<Solid::Device> _ret = Solid::Device::listFromQuery(*predicate, parentUdi_QString);
    // Convert QList<> from C++ memory to manually-managed C memory
    Solid__Device** _arr = static_cast<Solid__Device**>(malloc(sizeof(Solid__Device*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new Solid::Device(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of Solid__Device* */ Solid__Device_ListFromQuery23(const libqt_string predicate, const libqt_string parentUdi) {
    QString predicate_QString = QString::fromUtf8(predicate.data, predicate.len);
    QString parentUdi_QString = QString::fromUtf8(parentUdi.data, parentUdi.len);
    QList<Solid::Device> _ret = Solid::Device::listFromQuery(predicate_QString, parentUdi_QString);
    // Convert QList<> from C++ memory to manually-managed C memory
    Solid__Device** _arr = static_cast<Solid__Device**>(malloc(sizeof(Solid__Device*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new Solid::Device(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void Solid__Device_Delete(Solid__Device* self) {
    delete self;
}

File: include_original/extras-solid/libdevice.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const deviceinterface_enums = @import("libdeviceinterface.zig").enums;
const std = @import("std");

/// https://api.kde.org/solid-device.html
pub const solid__device = struct {
    /// New constructs a new Solid::Device object.
    ///
    ///
    pub fn New() QtC.Solid__Device {
        return qtc.Solid__Device_new();
    }

    /// New2 constructs a new Solid::Device object.
    ///
    /// ``` device: QtC.Solid__Device ```
    pub fn New2(device: ?*anyopaque) QtC.Solid__Device {
        return qtc.Solid__Device_new2(@ptrCast(device));
    }

    /// New3 constructs a new Solid::Device object.
    ///
    /// ``` udi: []const u8 ```
    pub fn New3(udi: []const u8) QtC.Solid__Device {
        const udi_str = qtc.libqt_string{
            .len = udi.len,
            .data = udi.ptr,
        };

        return qtc.Solid__Device_new3(udi_str);
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#allDevices)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn AllDevices(allocator: std.mem.Allocator) []QtC.Solid__Device {
        const _arr: qtc.libqt_list = qtc.Solid__Device_AllDevices();
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.Solid__Device, _arr.len) catch @panic("solid::device.AllDevices: Memory allocation failed");
        const _data: [*]QtC.Solid__Device = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#listFromType)
    ///
    /// ``` typeVal: *const deviceinterface_enums.Type, allocator: std.mem.Allocator ```
    pub fn ListFromType(typeVal: *const i32, allocator: std.mem.Allocator) []QtC.Solid__Device {
        const _arr: qtc.libqt_list = qtc.Solid__Device_ListFromType(@ptrCast(typeVal));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.Solid__Device, _arr.len) catch @panic("solid::device.ListFromType: Memory allocation failed");
        const _data: [*]QtC.Solid__Device = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#listFromQuery)
    ///
    /// ``` predicate: QtC.Solid__Predicate, allocator: std.mem.Allocator ```
    pub fn ListFromQuery(predicate: ?*anyopaque, allocator: std.mem.Allocator) []QtC.Solid__Device {
        const _arr: qtc.libqt_list = qtc.Solid__Device_ListFromQuery(@ptrCast(predicate));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.Solid__Device, _arr.len) catch @panic("solid::device.ListFromQuery: Memory allocation failed");
        const _data: [*]QtC.Solid__Device = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#listFromQuery)
    ///
    /// ``` predicate: []const u8, allocator: std.mem.Allocator ```
    pub fn ListFromQuery2(predicate: []const u8, allocator: std.mem.Allocator) []QtC.Solid__Device {
        const predicate_str = qtc.libqt_string{
            .len = predicate.len,
            .data = predicate.ptr,
        };
        const _arr: qtc.libqt_list = qtc.Solid__Device_ListFromQuery2(predicate_str);
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.Solid__Device, _arr.len) catch @panic("solid::device.ListFromQuery2: Memory allocation failed");
        const _data: [*]QtC.Solid__Device = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#storageAccessFromPath)
    ///
    /// ``` path: []const u8 ```
    pub fn StorageAccessFromPath(path: []const u8) QtC.Solid__Device {
        const path_str = qtc.libqt_string{
            .len = path.len,
            .data = path.ptr,
        };
        return qtc.Solid__Device_StorageAccessFromPath(path_str);
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#operator-eq)
    ///
    /// ``` self: QtC.Solid__Device, device: QtC.Solid__Device ```
    pub fn OperatorAssign(self: ?*anyopaque, device: ?*anyopaque) void {
        qtc.Solid__Device_OperatorAssign(@ptrCast(self), @ptrCast(device));
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#isValid)
    ///
    /// ``` self: QtC.Solid__Device ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.Solid__Device_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#udi)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn Udi(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Solid__Device_Udi(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("solid::device.Udi: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#parentUdi)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn ParentUdi(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Solid__Device_ParentUdi(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("solid::device.ParentUdi: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#parent)
    ///
    /// ``` self: QtC.Solid__Device ```
    pub fn Parent(self: ?*anyopaque) QtC.Solid__Device {
        return qtc.Solid__Device_Parent(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#vendor)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn Vendor(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Solid__Device_Vendor(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("solid::device.Vendor: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#product)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn Product(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Solid__Device_Product(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("solid::device.Product: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#icon)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn Icon(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Solid__Device_Icon(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("solid::device.Icon: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#emblems)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn Emblems(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.Solid__Device_Emblems(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("solid::device.Emblems: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("solid::device.Emblems: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#displayName)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn DisplayName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Solid__Device_DisplayName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("solid::device.DisplayName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#description)
    ///
    /// ``` self: QtC.Solid__Device, allocator: std.mem.Allocator ```
    pub fn Description(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Solid__Device_Description(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("solid::device.Description: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#isDeviceInterface)
    ///
    /// ``` self: QtC.Solid__Device, typeVal: *const deviceinterface_enums.Type ```
    pub fn IsDeviceInterface(self: ?*anyopaque, typeVal: *const i32) bool {
        return qtc.Solid__Device_IsDeviceInterface(@ptrCast(self), @ptrCast(typeVal));
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#asDeviceInterface)
    ///
    /// ``` self: QtC.Solid__Device, typeVal: *const deviceinterface_enums.Type ```
    pub fn AsDeviceInterface(self: ?*anyopaque, typeVal: *const i32) QtC.Solid__DeviceInterface {
        return qtc.Solid__Device_AsDeviceInterface(@ptrCast(self), @ptrCast(typeVal));
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#asDeviceInterface)
    ///
    /// ``` self: QtC.Solid__Device, typeVal: *const deviceinterface_enums.Type ```
    pub fn AsDeviceInterface2(self: ?*anyopaque, typeVal: *const i32) QtC.Solid__DeviceInterface {
        return qtc.Solid__Device_AsDeviceInterface2(@ptrCast(self), @ptrCast(typeVal));
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#listFromType)
    ///
    /// ``` typeVal: *const deviceinterface_enums.Type, parentUdi: []const u8, allocator: std.mem.Allocator ```
    pub fn ListFromType2(typeVal: *const i32, parentUdi: []const u8, allocator: std.mem.Allocator) []QtC.Solid__Device {
        const parentUdi_str = qtc.libqt_string{
            .len = parentUdi.len,
            .data = parentUdi.ptr,
        };
        const _arr: qtc.libqt_list = qtc.Solid__Device_ListFromType2(@ptrCast(typeVal), parentUdi_str);
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.Solid__Device, _arr.len) catch @panic("solid::device.ListFromType2: Memory allocation failed");
        const _data: [*]QtC.Solid__Device = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#listFromQuery)
    ///
    /// ``` predicate: QtC.Solid__Predicate, parentUdi: []const u8, allocator: std.mem.Allocator ```
    pub fn ListFromQuery22(predicate: ?*anyopaque, parentUdi: []const u8, allocator: std.mem.Allocator) []QtC.Solid__Device {
        const parentUdi_str = qtc.libqt_string{
            .len = parentUdi.len,
            .data = parentUdi.ptr,
        };
        const _arr: qtc.libqt_list = qtc.Solid__Device_ListFromQuery22(@ptrCast(predicate), parentUdi_str);
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.Solid__Device, _arr.len) catch @panic("solid::device.ListFromQuery22: Memory allocation failed");
        const _data: [*]QtC.Solid__Device = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/solid-device.html#listFromQuery)
    ///
    /// ``` predicate: []const u8, parentUdi: []const u8, allocator: std.mem.Allocator ```
    pub fn ListFromQuery23(predicate: []const u8, parentUdi: []const u8, allocator: std.mem.Allocator) []QtC.Solid__Device {
        const predicate_str = qtc.libqt_string{
            .len = predicate.len,
            .data = predicate.ptr,
        };
        const parentUdi_str = qtc.libqt_string{
            .len = parentUdi.len,
            .data = parentUdi.ptr,
        };
        const _arr: qtc.libqt_list = qtc.Solid__Device_ListFromQuery23(predicate_str, parentUdi_str);
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.Solid__Device, _arr.len) catch @panic("solid::device.ListFromQuery23: Memory allocation failed");
        const _data: [*]QtC.Solid__Device = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.Solid__Device ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.Solid__Device_Delete(@ptrCast(self));
    }
};
